pkg/services/dashboards/service/k8s_client.go

```diff
@@ -0,0 +1,125 @@
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/otel/attribute"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	dashboardv1alpha1 "github.com/grafana/grafana/apps/dashboard/pkg/apis/dashboard/v1alpha1"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/infra/tracing"
+	"github.com/grafana/grafana/pkg/services/apiserver"
+	"github.com/grafana/grafana/pkg/services/apiserver/client"
+	"github.com/grafana/grafana/pkg/services/apiserver/endpoints/request"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/search/sort"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/storage/legacysql/dualwrite"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+)
+
+type k8sHandlerFactory func(version string) client.K8sHandler
+
+type k8sHandlerWithFallback struct {
+	client.K8sHandler
+
+	newClientFunc k8sHandlerFactory
+	log           log.Logger
+}
+
+func NewK8sHandlerWithFallback(
+	cfg *setting.Cfg,
+	restConfigProvider apiserver.RestConfigProvider,
+	dashboardStore dashboards.Store,
+	userService user.Service,
+	resourceClient resource.ResourceClient,
+	sorter sort.Service,
+	dual dualwrite.Service,
+	parentLogger log.Logger,
+) client.K8sHandler {
+	newClientFunc := newK8sHandlerFactory(
+		cfg,
+		restConfigProvider,
+		dashboardStore,
+		userService,
+		resourceClient,
+		sorter,
+		dual,
+	)
+	return &k8sHandlerWithFallback{
+		K8sHandler:    newClientFunc(dashboardv1alpha1.VERSION),
+		newClientFunc: newClientFunc,
```

Comment:
Small nit (can be done in a follow-up PR if you agree with the suggestion) but we could cache clients per version to avoid extra initialization every time we need a different version.

Reply:
cache added in https://github.com/grafana/grafana/pull/103404/commits/85fd226a82587af4794b91f2086f420ccd8778ec